core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java

```diff
@@ -0,0 +1,114 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.storage.relational.utils;
+
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.gravitino.MetadataObject;
+import org.apache.gravitino.authorization.SecurableObject;
+import org.apache.gravitino.storage.relational.po.GroupPO;
+import org.apache.gravitino.storage.relational.po.RolePO;
+import org.apache.gravitino.storage.relational.po.SecurableObjectPO;
+import org.apache.gravitino.storage.relational.po.UserPO;
+import org.apache.gravitino.storage.relational.service.MetadataObjectService;
+import org.apache.gravitino.storage.relational.service.RoleMetaService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/* This class is a utilization class for creating kinds of suppliers */
+public class SupplierUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SupplierUtils.class);
+
+  private SupplierUtils() {}
+
+  public static Supplier<List<RolePO>> createRolePOsSupplier(UserPO userPO) {
+    return new Supplier<List<RolePO>>() {
+      private List<RolePO> rolePOs;
+      private boolean waitToLoad = true;
+
+      @Override
+      public List<RolePO> get() {
+        if (waitToLoad) {
+          rolePOs = RoleMetaService.getInstance().listRolesByUserId(userPO.getUserId());
+          waitToLoad = false;
+        }
+
+        return rolePOs;
+      }
+    };
+  }
+
+  public static Supplier<List<RolePO>> createRolePOsSupplier(GroupPO groupPO) {
+    return new Supplier<List<RolePO>>() {
+      private List<RolePO> rolePOS;
+      private boolean waitToLoad = true;
+
+      @Override
+      public List<RolePO> get() {
+        if (waitToLoad) {
+          rolePOS = RoleMetaService.getInstance().listRolesByGroupId(groupPO.getGroupId());
+          waitToLoad = false;
+        }
+
+        return rolePOS;
+      }
+    };
+  }
+
+  public static Supplier<List<SecurableObject>> createSecurableObjectsSupplier(RolePO rolePO) {
+    return new Supplier<List<SecurableObject>>() {
+      private List<SecurableObject> securableObjects;
+
+      private boolean waitToLoad = true;
+
+      @Override
+      public List<SecurableObject> get() {
+        if (waitToLoad) {
+          List<SecurableObjectPO> securableObjectPOs =
+              RoleMetaService.getInstance().listSecurableObjectsByRoleId(rolePO.getRoleId());
+
+          securableObjects = Lists.newArrayList();
+
+          for (SecurableObjectPO securableObjectPO : securableObjectPOs) {
+            String fullName =
+                MetadataObjectService.getMetadataObjectFullName(
+                    securableObjectPO.getType(), securableObjectPO.getMetadataObjectId());
+            if (fullName != null) {
+              securableObjects.add(
```

Comment:
I believe you need to clear `securableObjects` before you try to load them from storage, `securableObjects` is not empty when `get` fails halfway.

Reply:
You can see the line 88. I have cleared it.